tests/net/lib/http_server/core/src/main.c

```diff
@@ -46,183 +48,265 @@ static const unsigned char request_get_2_streams[] = {
+	zassert_equal(frame.stream_identifier, stream_id,
+		      "Invalid headers frame stream ID");
+	zassert_equal(frame.flags, flags, "Unexpected flags received");
 
-		zassert_true(type == 0x4, "Expected a SETTINGS frame");
-	}
+	/* Consume headers payload */
+	test_read_data(offset, frame.length);
+	test_consume_data(offset, frame.length);
+}
+
+static void expect_http2_data_frame(size_t *offset, int stream_id,
+				    const uint8_t *payload, size_t payload_len,
+				    uint8_t flags)
+{
+	struct http2_frame frame;
+
+	test_get_frame_header(offset, &frame);
+
+	zassert_equal(frame.type, HTTP2_DATA_FRAME, "Expected data frame");
+	zassert_equal(frame.stream_identifier, stream_id,
+		      "Invalid data frame stream ID");
+	zassert_equal(frame.flags, flags, "Unexpected flags received");
+	zassert_equal(frame.length, payload_len, "Unexpected data frame length");
+
+	/* Verify data payload */
+	test_read_data(offset, frame.length);
+	zassert_mem_equal(buf, payload, payload_len, "Unexpected data payload");
+	test_consume_data(offset, frame.length);
 }
 
-ZTEST(server_function_tests, test_http_upgrade)
+ZTEST(server_function_tests, test_http2_get_concurrent_streams)
 {
-	test_common(SUPPORT_HTTP_SERVER_UPGRADE);
+	size_t offset = 0;
+	int ret;
+
+	ret = zsock_send(client_fd, request_get_2_streams,
+			 sizeof(request_get_2_streams), 0);
+	zassert_not_equal(ret, -1, "send() failed (%d)", errno);
+
+	memset(buf, 0, sizeof(buf));
+
+	/* Settings frame is expected twice (server settings + settings ACK) */
+	expect_http2_settings_frame(&offset, false);
+	expect_http2_settings_frame(&offset, true);
+	expect_http2_headers_frame(&offset, TEST_STREAM_ID_1,
+				   HTTP2_FLAG_END_HEADERS);
+	expect_http2_data_frame(&offset, TEST_STREAM_ID_1, TEST_STATIC_PAYLOAD,
+				strlen(TEST_STATIC_PAYLOAD),
+				HTTP2_FLAG_END_STREAM);
+	expect_http2_headers_frame(&offset, TEST_STREAM_ID_2,
+				   HTTP2_FLAG_END_HEADERS);
+	expect_http2_data_frame(&offset, TEST_STREAM_ID_2, NULL, 0,
+				HTTP2_FLAG_END_STREAM);
 }
 
-ZTEST(server_function_tests, test_backward_compatibility)
+ZTEST(server_function_tests, test_http2_static_get)
 {
-	test_common(SUPPORT_BACKWARD_COMPATIBILITY);
+	size_t offset = 0;
+	int ret;
+
+	ret = zsock_send(client_fd, request_get_static_simple,
+			 sizeof(request_get_static_simple), 0);
+	zassert_not_equal(ret, -1, "send() failed (%d)", errno);
+
+	memset(buf, 0, sizeof(buf));
+
+	expect_http2_settings_frame(&offset, false);
+	expect_http2_settings_frame(&offset, true);
+	expect_http2_headers_frame(&offset, TEST_STREAM_ID_1,
+				   HTTP2_FLAG_END_HEADERS);
+	expect_http2_data_frame(&offset, TEST_STREAM_ID_1, TEST_STATIC_PAYLOAD,
+				strlen(TEST_STATIC_PAYLOAD),
+				HTTP2_FLAG_END_STREAM);
+}
+
+ZTEST(server_function_tests, test_http1_static_upgrade_get)
+{
+	static const char http1_request[] =
+		"GET / HTTP/1.1\r\n"
+		"Host: 127.0.0.1:8080\r\n"
+		"Accept: */*\r\n"
+		"Accept-Encoding: deflate, gzip, br\r\n"
```

Comment:
Just wondering what should we do in server code if the client is sending encoding list that the server is not supporting. Probably some enhancements are needed in server side in future PRs for handling this.

Reply:
I guess in embedded world there's not much to do, we can either reply with `406 Not Acceptable` or ignore the request. Nevertheless, we should indeed implement some form of content negotiation in the future.